### PR TITLE
feat: add support for embedded wikilink

### LINKED ISSUE #17: feature: support for images referred to by wikilink
Aside from documents, we might also want to refer to images using wikilink e.g. `![[some-picture.png]]`

### PR DESCRIPTION
**Pull request by benclmnt**
Closes #17

### CODE REVIEW

File: ast.go [hunk 1]
```diff
@@ -35,6 +35,10 @@ type Node struct {
 	// For links in the form, [[Foo bar#Baz qux]], this is the portion
 	// after the "#".
 	Fragment []byte
+
+	// True if it starts with a bang (!).
+	// This indicates that the resource should be embedded (e.g. images)
+	IsEmbed bool
```
Comment (abhinav) [inline ast.go:41]: Can we rename this just 'Embed'?

File: parser.go [hunk 2]
```diff
@@ -49,9 +49,16 @@ func (p *Parser) Trigger() []byte {
...
+// A wikilink can also be embedded, i.e. starts with a bang (!).
+// A few examples of embedded wikilink:
+//
+//  ![[My page]] (simple)
+//  ![[My page#fragment|click here]]
 func (p *Parser) Parse(_ ast.Node, block text.Reader, _ parser.Context) ast.Node {
 	line, seg := block.PeekLine()
-	if !bytes.HasPrefix(line, _open) {
+	start := bytes.Index(line, _open)
+	if start > 1 || (start == 1 && line[0] != '!') {
```
Comment (abhinav) [inline parser.go:61]: I think we can make this clearer with something like:

```
line, seg := ..
var embed bool
switch {
case bytes.HasPrefix(line, _open):
  // not an embed
case bytes.HasPrefix(line, _embedOpen):
  embed = true
default:
  return nil
}
```

Where `_embedOpen = []byte("![[")`.
It makes what we're searching for slightly more obvious without offset munging.

**Review by abhinav — COMMENTED** [review]
Thanks for the PR, @benclmnt!
I agree that it makes sense to add the `![[..]]` syntax,
and that making that a property of Node makes sense.

I think we clarify in the doc for the Embed field
that we don't actually do the embedding,
and that it's treated like any other link.

Lastly, would you consider adding some tests for this new behavior?

**Comment by benclmnt** [comment]
Hi @abhinav thanks for reviewing the code. I have made the changes as you requested.

However, I think the behaviour of the default resolver / renderer should be changed as well. 
- The default resolver should only add `.html` prefix on files without extension
- For the renderer, do you think it would be better to use the original to-html renderer in goldmark. Instead, we can transform the ast nodes using NodeTransformer before passing to the renderer.

A proof-of-concept of the approach:
```go
type WikilinkTransformer struct{}

func (w *WikilinkTransformer) Transform(node *ast.Document, reader text.Reader, pc parser.Context) {
	source := reader.Source()

	err := ast.Walk(node, func(node ast.Node, entering bool) (ast.WalkStatus, error) {
		// Each node will be visited twice, once when it is first encountered (entering), and again
		// after all the node's children have been visited (if any). Skip the latter.
		if !entering {
			return ast.WalkContinue, nil
		}

		if node.Kind() == wikilink.Kind {
			wikilinkNode := node.(*wikilink.Node)
			w.Rewrite(wikilinkNode, source)
		}

		return ast.WalkContinue, nil
	})

	if err != nil {
		log.Fatal("Error encountered while transforming AST:", err)
	}
}

// Rewrite wikilinks into normal markdown links
func (w *WikilinkTransformer) Rewrite(node *wikilink.Node, source []byte) {
	parent := node.Parent()

	link := ast.NewLink()
	// Link has a single child: the text to display
	link.AppendChild(link, node.FirstChild())
	href := resolver.ResolveWikilink() // <-- put logic here.
	link.Destination = href

	nextSibling := node.NextSibling()
	if node.IsEmbed {
		img := ast.NewImage(link)
		parent.ReplaceChild(parent, node, img)
	} else {
		parent.ReplaceChild(parent, node, link)
	}
	// since we replaced the node, node.next is now set to nil.
	// so that we can still traverse the siblings, we need to set it back
	node.SetNextSibling(nextSibling)
}
```

To use:
```go
md := goldmark.New(goldmark.WithParserOptions(
			parser.WithASTTransformers(util.Prioritized(&WikilinkTransformer{}, 20))
	),
)
```

**Comment by abhinav** [comment]
Good point on the rendering logic changes.

So, part of the reason I didn't just transform wikilink.Node to ast.Link was that I wanted the raw wikilink node accessible in the parsed AST to be able to differentiate between when someone used `[[Foo]]` and `[Foo](...)`.
I'd like the possibility for someone to be able to override the rendering logic for just wikilinks (like say, add an extra class to all wikilinks) without changing the logic for regular Link nodes. Transforming them into link nodes loses information and makes it difficult to render wikilinks differently from normal links.

With embedding, it gets a bit more complicated than just images. e.g.: Obsidian supports [embedding a number of things](https://help.obsidian.md/How+to/Embed+files), including pages (when you do `![[Foo]]`, and it's another note, not an image, Obsidian includes the contents of that page).

The logic for what can be embedded and how is too use-case-specific to hard-code into this library.
Perhaps we can introduce an Embedder interface that lets users specify how to embed a specific thing?

```go
type Embedder interface {
  EmbedWikilink(*Node) (n ast.Node, err error)
}

var ErrUnsupportedEmbed = errors.New("embed not supported")
// If the Embedder doesn't know how to embed this node, it can return ErrUnsupportedEmbed.

// EmbedTransformer transforms wikilink embeds into their embedded content.
type EmbedTransformer struct {
  Embedder Embedder
}

type Extender struct {
  // ...

  // EmbedTransformer is installed only if this field is set.
  Embedder Embedder
}
```

We can include a sample `ImageEmbedder` implementation based on your logic above,
that they can pass in as the Embedder, or model after for custom embedding logic.
Extender will install the EmbedTransformer and Parser will support `![[..]]` links only if an Embedder was specified.

One thing I'm unsure about that I'd love input on is: should this be a separate EmbedNode or an `Embed bool` field on Node? If we make it a separate Node type, it'll ensure that the EmbedNode never makes it to the Renderer.

What do you think?

**Comment by benclmnt** [comment]
> The logic for what can be embedded and how is too use-case-specific to hard-code into this library.

agree!

> Perhaps we can introduce an Embedder interface that lets users specify how to embed a specific thing?

I think users of this library would still need to override the renderer to achieve the embed functionality for files / notes. so imho the default renderer should just support the image usecase because `label` is in the official commonmark spec 

I think a separate EmbedNode might offer more flexibility than an `Embed bool` but given the above considerations, I don't think we need to separate it now unless we do have a concrete usecase that requires that.

**Comment by abhinav** [comment]
Sorry, just to clarify:

> so imho the default renderer should just support the image usecase because `label` is in the official commonmark spec

By that, do you mean that we should add the Embedder interface, but have the extension support image embedding regardless of whether an Embedder is specified or not? Or defer adding an Embedder interface until later and only do image embedding for now? Either sounds reasonable.

> I think a separate EmbedNode might offer more flexibility than an Embed bool but given the above considerations, I don't think we need to separate it now unless we do have a concrete usecase that requires that.

Fair. I lean towards the second type but I don't feel strongly about it so I'll defer to you.

**Comment by abhinav** [comment]
FWIW, the current version of the PR looks good to me as-is.
I'm happy to merge this if you'd like to follow up with rendering separately,
or if you think we should let users figure out the rendering of embeds themselves for now.

**Comment by benclmnt** [comment]
> By that, do you mean that we should add the Embedder interface, but have the extension support image embedding regardless of whether an Embedder is specified or not? Or defer adding an Embedder interface until later and only do image embedding for now? Either sounds reasonable.

I think we should defer adding an Embedder interface until later, and only do image embedding for now. I'll submit in a separate PR.

Thanks for reviewing this and happy new year! 🥳

**Comment by abhinav** [comment]
Sounds good!
Thank you, and happy new year to you as well!

**Pull request merged**